util/dvsim/Testplan.py

```diff
@@ -0,0 +1,686 @@
+                                    f"  Tests: {self.tests}\n")
+
+    def _validate(self):
+        super()._validate()
+        if self.milestone not in Testpoint.milestones:
+            raise ValueError(f"Testpoint milestone {self.milestone} is "
+                             f"invalid:\n{self}\nLegal values: "
+                             f"Testpoint.milestones")
+
+    def do_substitutions(self, substitutions):
+        '''Substitute {wildcards} in tests
+
+        If tests have {wildcards}, they are substituted with the 'correct'
+        values using the key=value pairs provided by the substitutions arg. If
+        wildcards are present but no replacement is available, then the
+        wildcards are replaced with an empty string.
+
+        substitutions is a dict of key-value pairs.
+        '''
+        resolved_tests = []
+        for test in self.tests:
+            match = re.findall(r"{([A-Za-z0-9\_]+)}", test)
+            if not match:
+                resolved_tests.append(test)
+                continue
+
+            # 'match' is a list of wildcards used in the test. Get their
+            # corresponding values.
+            subst = {item: substitutions.get(item, "") for item in match}
+
+            resolved = [test]
+            for item, value in subst.items():
+                if isinstance(value, list):
+                    for v in value:
+                        resolved += [
+                            t.replace("{" + item + "}", v) for t in resolved
+                        ]
+                else:
+                    resolved = [
+                        t.replace("{" + item + "}", value) for t in resolved
+                    ]
+
+            resolved_tests.extend(resolved)
+
+        self.tests = resolved_tests
+
+    def map_test_results(self, test_results):
+        """Map test results to tests against this testpoint.
+
+        Given a list of test results find the ones that match the tests listed
+        in this testpoint and buiild a structure. If no match is found, or if
+        self.tests is an empty list, indicate 0/1 passing so that it is
+        factored into the final total.
+        """
+        for tr in test_results:
+            assert isinstance(tr, TestResult)
+            if tr.name in self.tests:
+                tr.mapped = True
+                self.test_results.append(tr)
+
+        # Did we map all tests in this testpoint? If we are mapping the full
+        # testplan, then count the ones not found as "not run", i.e. 0 / 0.
+        tests_mapped = [tr.name for tr in self.test_results]
+        for test in self.tests:
+            if test not in tests_mapped:
+                self.test_results.append(TestResult(test))
+
+        # If no written tests were indicated for this testpoint, then reuse
+        # the testpoint name to count towards "not run".
+        if not self.tests:
+            self.test_results = [TestResult(self.name)]
```

Comment:
Ah... this is where you use the TestResult constructor with default values. I think it would be clearer to provide the 'zeroes' explicitly.